File: backend/taskwarrior/tasks.go
package taskwarrior

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"
	"time"
)

// taskDateLayout is Taskwarrior's export format: ISO 8601 basic, always UTC.
// It is not what time.RFC3339 parses, so every date field is converted before
// it reaches the browser.
const taskDateLayout = "20060102T150405Z"

// StatusFilter selects which tasks to export.
type StatusFilter string

const (
	StatusPending   StatusFilter = "pending"
	StatusCompleted StatusFilter = "completed"
	StatusAll       StatusFilter = "all"
)

// ParseStatusFilter validates a filter coming from a query parameter.
// Only known values are accepted; the value is never passed through to the
// `task` command line as free text.
func ParseStatusFilter(raw string) (StatusFilter, error) {
	switch StatusFilter(strings.TrimSpace(raw)) {
	case "", StatusPending:
		return StatusPending, nil
	case StatusCompleted:
		return StatusCompleted, nil
	case StatusAll:
		return StatusAll, nil
	default:
		return "", fmt.Errorf("taskwarrior: unknown status filter %q", raw)
	}
}

// args returns the fixed command-line filter for this selection.
func (f StatusFilter) args() []string {
	switch f {
	case StatusCompleted:
		return []string{"status:completed"}
	case StatusAll:
		return nil
	default:
		return []string{"status:pending"}
	}
}

// Annotation is a timestamped note attached to a task.
type Annotation struct {
	Entry       string `json:"entry"`
	Description string `json:"description"`
}

// Task is a single task as presented to the UI.
type Task struct {
	// ID is Taskwarrior's short working id. It is only assigned to pending
	// tasks; completed tasks export as 0, so UUID is the stable identity.
	ID          int          `json:"id"`
	UUID        string       `json:"uuid"`
	Description string       `json:"description"`
	Status      string       `json:"status"`
	Project     string       `json:"project,omitempty"`
	Tags        []string     `json:"tags,omitempty"`
	Priority    string       `json:"priority,omitempty"`
	Urgency     float64      `json:"urgency"`
	Due         string       `json:"due,omitempty"`
	Scheduled   string       `json:"scheduled,omitempty"`
	Wait        string       `json:"wait,omitempty"`
	Entry       string       `json:"entry,omitempty"`
	Start       string       `json:"start,omitempty"`
	End         string       `json:"end,omitempty"`
	Modified    string       `json:"modified,omitempty"`
	Recur       string       `json:"recur,omitempty"`
	Depends     []string     `json:"depends,omitempty"`
	Annotations []Annotation `json:"annotations,omitempty"`
	// UDAs holds user-defined attributes, which differ per installation and
	// would be silently dropped by a fixed schema.
	UDAs map[string]string `json:"udas,omitempty"`
}

// TaskList is the payload served by the tasks endpoint.
type TaskList struct {
	Status StatusFilter `json:"status"`
	Counts Counts       `json:"counts"`
	Tasks  []Task       `json:"tasks"`
	// UDALabels maps a UDA name to its configured human-readable label.
	UDALabels map[string]string `json:"udaLabels,omitempty"`
}

// Counts summarises the returned set by status.
type Counts struct {
	Total     int `json:"total"`
	Pending   int `json:"pending"`
	Completed int `json:"completed"`
	Waiting   int `json:"waiting"`
	Recurring int `json:"recurring"`
	Deleted   int `json:"deleted"`
	Active    int `json:"active"`
	Overdue   int `json:"overdue"`
}

// knownFields are the attributes mapped onto Task. Anything else that
// Taskwarrior exports is treated as a user-defined attribute.
var knownFields = map[string]bool{
	"id": true, "uuid": true, "description": true, "status": true,
	"project": true, "tags": true, "priority": true, "urgency": true,
	"due": true, "scheduled": true, "wait": true, "entry": true,
	"start": true, "end": true, "modified": true, "recur": true,
	"depends": true, "annotations": true,
	// Recurrence bookkeeping that is meaningless in a task list.
	"rtype": true, "mask": true, "imask": true, "parent": true,
}

// parseTaskDate converts a Taskwarrior timestamp to RFC 3339 so the browser
// can parse it. Unparseable values are passed through unchanged rather than
// discarded, so a surprising format stays visible instead of vanishing.
func parseTaskDate(raw string) string {
	if raw == "" {
		return ""
	}
	parsed, err := time.Parse(taskDateLayout, raw)
	if err != nil {
		return raw
	}
	return parsed.UTC().Format(time.RFC3339)
}

// parseTasks decodes `task export` output.
func parseTasks(data []byte) ([]Task, error) {
	var raw []map[string]json.RawMessage
	if err := json.Unmarshal(data, &raw); err != nil {
		return nil, fmt.Errorf("taskwarrior: decoding task export: %w", err)
	}

	tasks := make([]Task, 0, len(raw))
	for _, entry := range raw {
		tasks = append(tasks, decodeTask(entry))
	}
	return tasks, nil
}

func decodeTask(entry map[string]json.RawMessage) Task {
	task := Task{
		ID:          decodeInt(entry["id"]),
		UUID:        decodeString(entry["uuid"]),
		Description: decodeString(entry["description"]),
		Status:      decodeString(entry["status"]),
		Project:     decodeString(entry["project"]),
		Tags:        decodeStrings(entry["tags"]),
		Priority:    decodeString(entry["priority"]),
		Urgency:     decodeFloat(entry["urgency"]),
		Due:         parseTaskDate(decodeString(entry["due"])),
		Scheduled:   parseTaskDate(decodeString(entry["scheduled"])),
		Wait:        parseTaskDate(decodeString(entry["wait"])),
		Entry:       parseTaskDate(decodeString(entry["entry"])),
		Start:       parseTaskDate(decodeString(entry["start"])),
		End:         parseTaskDate(decodeString(entry["end"])),
		Modified:    parseTaskDate(decodeString(entry["modified"])),
		Recur:       decodeString(entry["recur"]),
		Depends:     decodeStrings(entry["depends"]),
		Annotations: decodeAnnotations(entry["annotations"]),
	}

	for name, value := range entry {
		if knownFields[name] {
			continue
		}
		if task.UDAs == nil {
			task.UDAs = make(map[string]string)
		}
		task.UDAs[name] = decodeScalar(value)
	}

	return task
}

func decodeAnnotations(raw json.RawMessage) []Annotation {
	if len(raw) == 0 {
		return nil
	}
	var annotations []Annotation
	if err := json.Unmarshal(raw, &annotations); err != nil {
		return nil
	}
	for i := range annotations {
		annotations[i].Entry = parseTaskDate(annotations[i].Entry)
	}
	return annotations
}

func countTasks(tasks []Task, now time.Time) Counts {
	var counts Counts
	counts.Total = len(tasks)
	for _, task := range tasks {
		switch task.Status {
		case "pending":
			counts.Pending++
		case "completed":
			counts.Completed++
		case "waiting":
			counts.Waiting++
		case "recurring":
			counts.Recurring++
		case "deleted":
			counts.Deleted++
		}
		if task.Start != "" {
			counts.Active++
		}
		// A completed task with a past due date is not overdue any more.
		if task.Status == "pending" && task.Due != "" {
			if due, err := time.Parse(time.RFC3339, task.Due); err == nil && due.Before(now) {
				counts.Overdue++
			}
		}
	}
	return counts
}

// sortTasks orders by urgency, matching what `task list` shows first. UUID
// breaks ties so the order is stable across requests.
func sortTasks(tasks []Task) {
	sort.SliceStable(tasks, func(i, j int) bool {
		if tasks[i].Urgency != tasks[j].Urgency {
			return tasks[i].Urgency > tasks[j].Urgency
		}
		return tasks[i].UUID < tasks[j].UUID
	})
}

File: ui/src/taskwarrior/tasks.ts
import { ConfigError } from "./client";
import type { SortSpec } from "./sortspec";

/** Taskwarrior's export format: ISO 8601 basic, always UTC. It is not what
 *  `Date` parses, so every date field is converted before it is rendered. */
const taskDateFormat = /^(\d{4})(\d{2})(\d{2})T(\d{2})(\d{2})(\d{2})Z$/;

/** Selects which tasks to export. */
export type StatusFilter = "pending" | "completed" | "all";

/** Validates a filter coming from the URL. Only known values are accepted;
 *  the value is never passed through to the `task` command line as free
 *  text. */
export function parseStatusFilter(raw: string): StatusFilter {
  switch (raw.trim()) {
    case "":
    case "pending": return "pending";
    case "completed": return "completed";
    case "all": return "all";
    default:
      throw new ConfigError(
        `taskwarrior: unknown status filter "${raw}"`,
        "Supported values are pending, completed and all.",
      );
  }
}

/** The fixed command-line filter for a selection. */
export function statusFilterArgs(filter: StatusFilter): string[] {
  switch (filter) {
    case "completed": return ["status:completed"];
    case "all": return [];
    default: return ["status:pending"];
  }
}

/** A timestamped note attached to a task. */
export interface Annotation {
  entry: string;
  description: string;
}

export interface Task {
  /** Taskwarrior's short working id. Only pending tasks get one; completed
   *  tasks export as 0, so `uuid` is the stable identity. */
  id: number;
  uuid: string;
  description: string;
  status: string;
  project?: string;
  tags?: string[];
  priority?: string;
  urgency: number;
  due?: string;
  scheduled?: string;
  wait?: string;
  entry?: string;
  start?: string;
  end?: string;
  modified?: string;
  recur?: string;
  depends?: string[];
  annotations?: Annotation[];
  /** User-defined attributes, which differ per installation and would be
   *  silently dropped by a fixed schema. */
  udas?: Record<string, string>;
}

/** Summarises a task list by status. */
export interface Counts {
  total: number;
  pending: number;
  completed: number;
  waiting: number;
  recurring: number;
  deleted: number;
  active: number;
  overdue: number;
}

export interface TaskList {
  status: StatusFilter;
  counts: Counts;
  tasks: Task[];
  /** The order applied, so the interface can name it. */
  sort: SortSpec;
  /** Maps a UDA name to its configured human-readable label. */
  udaLabels?: Record<string, string>;
}

/** The attributes mapped onto Task. Anything else that Taskwarrior exports
 *  is treated as a user-defined attribute. */
const knownFields = new Set([
  "id", "uuid", "description", "status",
  "project", "tags", "priority", "urgency",
  "due", "scheduled", "wait", "entry",
  "start", "end", "modified", "recur",
  "depends", "annotations",
  // Recurrence bookkeeping that is meaningless in a task list.
  "rtype", "mask", "imask", "parent",
]);

/** Converts a Taskwarrior timestamp to RFC 3339 so it can be compared and
 *  formatted. Unparseable values are passed through unchanged rather than
 *  discarded, so a surprising format stays visible instead of vanishing. */
export function parseTaskDate(raw: string): string {
  if (raw === "") return "";

  const match = taskDateFormat.exec(raw);
  if (match === null) return raw;

  const [year = 0, month = 0, day = 0, hour = 0, minute = 0, second = 0] =
    match.slice(1).map(Number);
  const stamp = Date.UTC(year, month - 1, day, hour, minute, second);
  const parsed = new Date(stamp);
  // Date.UTC rolls an impossible date over into the next month rather than
  // rejecting it, so the round trip is what actually validates the input.
  if (
    parsed.getUTCFullYear() !== year || parsed.getUTCMonth() !== month - 1 ||
    parsed.getUTCDate() !== day || parsed.getUTCHours() !== hour ||
    parsed.getUTCMinutes() !== minute || parsed.getUTCSeconds() !== second
  ) {
    return raw;
  }
  // toISOString always carries milliseconds, which Taskwarrior never has.
  return parsed.toISOString().replace(".000Z", "Z");
}

// The decode helpers are deliberately forgiving: a task with one unexpected
// field type should still be listed, with that field empty, rather than
// failing the whole request.

function decodeString(raw: unknown): string {
  return typeof raw === "string" ? raw : "";
}

function decodeNumber(raw: unknown): number {
  return typeof raw === "number" && Number.isFinite(raw) ? raw : 0;
}

function decodeStrings(raw: unknown): string[] | undefined {
  if (Array.isArray(raw)) {
    if (!raw.every((entry) => typeof entry === "string")) return undefined;
    return raw.length === 0 ? undefined : (raw as string[]);
  }
  // Older Taskwarrior versions export dependencies as a comma-separated
  // string rather than a list.
  const single = decodeString(raw);
  return single === "" ? undefined : single.split(",");
}

function decodeAnnotations(raw: unknown): Annotation[] | undefined {
  if (!Array.isArray(raw) || raw.length === 0) return undefined;

  const annotations: Annotation[] = [];
  for (const entry of raw) {
    if (entry === null || typeof entry !== "object") return undefined;
    const record = entry as Record<string, unknown>;
    annotations.push({
      entry: parseTaskDate(decodeString(record.entry)),
      description: decodeString(record.description),
    });
  }
  return annotations;
}

/** Renders a user-defined attribute of unknown type as text. */
function decodeScalar(raw: unknown): string {
  switch (typeof raw) {
    case "string": return raw;
    case "boolean": return String(raw);
    // UDAs of type numeric are frequently whole numbers; 3 reads better
    // than 3.000000.
    case "number": return String(raw);
    case "undefined": return "";
    default:
      if (raw === null) return "";
      // Arrays and objects have no sensible flat rendering; keep the JSON.
      try {
        return JSON.stringify(raw) ?? "";
      } catch {
        return "";
      }
  }
}

/** Sets a field only when it carries something, so an absent attribute stays
 *  absent instead of becoming an empty string. */
function assign<T>(task: Task, key: keyof Task, value: T | undefined): void {
  if (value === undefined || value === "") return;
  (task as unknown as Record<string, unknown>)[key] = value;
}

function decodeTask(entry: Record<string, unknown>): Task {
  const task: Task = {
    id: decodeNumber(entry.id),
    uuid: decodeString(entry.uuid),
    description: decodeString(entry.description),
    status: decodeString(entry.status),
    urgency: decodeNumber(entry.urgency),
  };

  assign(task, "project", decodeString(entry.project));
  assign(task, "tags", decodeStrings(entry.tags));
  assign(task, "priority", decodeString(entry.priority));
  assign(task, "due", parseTaskDate(decodeString(entry.due)));
  assign(task, "scheduled", parseTaskDate(decodeString(entry.scheduled)));
  assign(task, "wait", parseTaskDate(decodeString(entry.wait)));
  assign(task, "entry", parseTaskDate(decodeString(entry.entry)));
  assign(task, "start", parseTaskDate(decodeString(entry.start)));
  assign(task, "end", parseTaskDate(decodeString(entry.end)));
  assign(task, "modified", parseTaskDate(decodeString(entry.modified)));
  assign(task, "recur", decodeString(entry.recur));
  assign(task, "depends", decodeStrings(entry.depends));
  assign(task, "annotations", decodeAnnotations(entry.annotations));

  const udas: Record<string, string> = {};
  for (const [name, value] of Object.entries(entry)) {
    if (knownFields.has(name)) continue;
    udas[name] = decodeScalar(value);
  }
  if (Object.keys(udas).length > 0) task.udas = udas;

  return task;
}

/** Decodes `task export` output. */
export function parseTasks(output: string): Task[] {
  let raw: unknown;
  try {
    raw = JSON.parse(output);
  } catch (error) {
    throw new ConfigError(
      `taskwarrior: decoding task export: ${(error as Error).message}`,
      "Run `task export` in a terminal to see what Taskwarrior printed.",
    );
  }
  if (!Array.isArray(raw)) {
    throw new ConfigError("taskwarrior: decoding task export: expected a list of tasks");
  }

  return raw.map((entry) =>
    decodeTask(
      entry !== null && typeof entry === "object"
        ? (entry as Record<string, unknown>)
        : {},
    ),
  );
}

export function countTasks(tasks: Task[], now: Date): Counts {
  const counts: Counts = {
    total: tasks.length,
    pending: 0, completed: 0, waiting: 0,
    recurring: 0, deleted: 0, active: 0, overdue: 0,
  };

  for (const task of tasks) {
    switch (task.status) {
      case "pending": counts.pending++; break;
      case "completed": counts.completed++; break;
      case "waiting": counts.waiting++; break;
      case "recurring": counts.recurring++; break;
      case "deleted": counts.deleted++; break;
    }
    if (task.start !== undefined) counts.active++;
    // A completed task with a past due date is not overdue any more.
    if (task.status === "pending" && task.due !== undefined) {
      const due = new Date(task.due);
      if (!Number.isNaN(due.valueOf()) && due < now) counts.overdue++;
    }
  }
  return counts;
}
